### ISSUE TITLE
Panic: makeslice: cap out of range

### ISSUE THREAD

**Issue by erizocosmico** [bblfsh/client-go#69]
Sometimes, client-go panics with the following error:

```
runtime error: makeslice: cap out of range
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/panic.go:502 (0x402be48)
	gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/slice.go:58 (0x4041d31)
	makeslice: panic(errorString("makeslice: cap out of range"))
/Users/erizocosmico/go/src/github.com/src-d/gitbase/vendor/gopkg.in/bblfsh/client-go.v2/tools/bindings.go:245 (0x47c53d7)
	com/src-d/gitbase/vendor/gopkg.in/bblfsh/client-go.v2/tools.getPropertyKeys: keys := make([]string, 0, len(p))
/Users/erizocosmico/go/src/github.com/src-d/gitbase/vendor/gopkg.in/bblfsh/client-go.v2/tools/bindings.go:256 (0x47c563a)
	com/src-d/gitbase/vendor/gopkg.in/bblfsh/client-go.v2/tools.goGetPropertyKey: keys := getPropertyKeys(ptr)
_cgo_gotypes.go:340 (0x47c4601)
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/asm_amd64.s:573 (0x4054f9a)
	call32: CALLFN(·call32, 32)
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/cgocall.go:316 (0x40042fb)
	cgocallbackg1: reflectcall(nil, unsafe.Pointer(cb.fn), cb.arg, uint32(cb.argsize), 0)
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/cgocall.go:194 (0x40040c9)
	cgocallbackg: cgocallbackg1(ctxt)
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/asm_amd64.s:826 (0x405656a)
	cgocallback_gofunc: CALL	runtime·cgocallbackg(SB)
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/asm_amd64.s:673 (0x4056401)
	asmcgocall: CALL	gosave<>(SB)
/usr/local/Cellar/go/1.10.1/libexec/src/runtime/cgocall.go:131 (0x4003f77)
	cgocall: errno := asmcgocall(fn, arg)
_cgo_gotypes.go:107 (0x47c3f09)
/Users/erizocosmico/go/src/github.com/src-d/gitbase/vendor/gopkg.in/bblfsh/client-go.v2/tools/bindings.go:122 (0x47c50c3)
	com/src-d/gitbase/vendor/gopkg.in/bblfsh/client-go.v2/tools.Filter: if !C.Filter(ptr, cquery) {
```

We're checking the errors and the response status is Ok.

If it's of any help, the XPath for the filter was `//FunctionGroup`.

**Comment by dennwc** [comment]
This is not the only error that you can get on that build. Seems like C does something nasty with Go pointers/memory.

Switching from `void*` to `int` (passing opaque handles instead of real Go pointers to C) fixes the problem for me. But it will introduce a memory leak, since Go will have to keep track the lifetime of these handles and this cannot be done automatically with current interface that `libuast` provides.

**Comment by juanjux** [comment]
@dennwc if this is for filter and not iterators, we could add to libuast an alternative filter version that takes a callback, like the Iterators have. With this callback the C part could do the memory tracking of passed pointers (that's how it's done in the Python and Scala clients if you check the code).

If this is related to Iterators, well, the callback already exists :)

**Comment by dennwc** [comment]
@juanjux The problem in leaking memory is that I want to change the client to keep the map of handles to nodes. But `NodeIface` has no way to know when `Uast` object is deallocated, so it cannot free map entries in the right time. Right now the solution I use is to wipe the map inside `Filter`, but it won't work for iterators.

**Comment by dennwc** [comment]
Should be fixed in [v2.8.0](https://github.com/bblfsh/client-go/releases/tag/v2.8.0).
